sdk/core/Azure.Core/src/Shared/AzureSasCredentialPolicy.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    internal class AzureSasCredentialPolicy : HttpPipelineSynchronousPolicy
+    {
+        private readonly AzureSasCredential _credential;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureSasCredentialPolicy"/> class.
+        /// </summary>
+        /// <param name="credential">The <see cref="AzureSasCredentialPolicy"/> used to authenticate requests.</param>
+        public AzureSasCredentialPolicy(AzureSasCredential credential)
+        {
+            Argument.AssertNotNull(credential, nameof(credential));
+            _credential = credential;
+        }
+
+        /// <inheritdoc/>
+        public override void OnSendingRequest(HttpMessage message)
+        {
+            base.OnSendingRequest(message);
+            string query = message.Request.Uri.Query;
+            string signature = _credential.Signature;
+            if (signature.StartsWith("?", StringComparison.InvariantCulture))
+            {
+                signature = signature.Substring(1);
+            }
+            if (!query.Contains(signature))
```

Comment:
Why is this check needed?  Won't we validate no SAS URI on the client .ctor?

Reply:
Can the `query` be null?

Reply:
It's not related to presence of the SAS in client's URI.
I used `AzureKeyCredentialPolicy` as a blueprint for this one and found [this test](https://github.com/Azure/azure-sdk-for-net/blob/1ee01cbe78969514ce104502fc9fa627e2c5a293/sdk/core/Azure.Core/tests/AzureKeyCredentialPolicyTests.cs#L30). That test checks idempotency of the policy if request is retried. I created similar tests for SAS policy and discovered that we'd be appending SAS on each retry. So this is to prevent that.

[Query](https://github.com/Azure/azure-sdk-for-net/blob/793c5a7577ab0c6057893ed57eb13bdfd25f0bcd/sdk/core/Azure.Core/src/RequestUriBuilder.cs#L75) is empty string in worst case.

Reply:
Java SDK handles this problem in an interesting way. Its retry policy makes a copy of its HttpMessage equivalent before calling rest of pipeline chain, so that retries won't step on each other. Maybe doing something like that in .NET would be right solution to this problem.